### PR TITLE
feat: partial delete form

### PR DESCRIPTION
**Pull request by DorisWongNHS**
- listen to the delete event queue to handle partial delete in DB
- add `deleteType` and `fixedFields` in S3 Metadata when new forms are saved
- Mongock migration to bulk add `deleteType` and `fixedFields` to Metadata of all the existing forms

TIS21-4009
TIS21-4019
TIS21-4077
TIS21-4023

### CODE REVIEW

File: src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          log.info("Partial delete successfully for trainee {} with form Id {} (FormRPartA)",
+              traineeTisId, formId);
+        } else if (eventDetails[2].equals("formr-b")) {
+          formRPartBService.partialDeleteFormRPartBById(formId, traineeTisId, fixFields);
+          log.info("Partial delete successfully for trainee {} with form Id {} (FormRPartB)",
+              traineeTisId, formId);
+        }
+      } else {
+        log.error("Unexpected deleteType of form: {}" + deleteEvent.getDeleteType());
+        throw new ApplicationException("Unexpected deleteType of form: "
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java:80]: Is the intention to send non-`PARTIAL` deletes to the DLQ for review?

File: src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartAServiceImpl.java [hunk 2]
```diff
@@ -110,4 +118,40 @@ public FormRPartADto getFormRPartAById(String id, String traineeTisId) {
...
+  public FormRPartADto partialDeleteFormRPartAById(
+      String id, String traineeTisId, String[] fixedFields) {
+    log.info("Request to partial delete FormRPartA by id : {}", id);
+
+    try {
+      FormRPartA formRPartA = repository.findByIdAndTraineeTisId(UUID.fromString(id), traineeTisId)
+          .orElse(null);
+
+      if (formRPartA != null) {
+        ObjectMapper objectMapper = new ObjectMapper().findAndRegisterModules();
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartAServiceImpl.java:134]: This should probably be injected via the constructor.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartAServiceImpl.java [hunk 3]
```diff
@@ -110,4 +118,40 @@ public FormRPartADto getFormRPartAById(String id, String traineeTisId) {
...
+        JsonNode jsonForm = objectMapper.valueToTree(formRPartA);
+
+        for (Iterator<String> fieldIterator = jsonForm.fieldNames(); fieldIterator.hasNext(); ) {
+          String fieldName = fieldIterator.next();
+
+          if (!Set.of(fixedFields).contains(fieldName)) {
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartAServiceImpl.java:140]: `Set.of(fixedFields)` here will create a new `Set` on every loop iteration, I recommend initializing the set outside of the loop. Alternatively, changing the method parameter to require a Set might be clearer.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartBServiceImpl.java [hunk 4]
```diff
@@ -105,4 +113,42 @@ public FormRPartBDto getFormRPartBById(String id, String traineeTisId) {
...
+          .orElse(null);
+
+      if (formRPartB != null) {
+        ObjectMapper objectMapper = new ObjectMapper().findAndRegisterModules();
+        JsonNode jsonForm = objectMapper.valueToTree(formRPartB);
+
+        for (Iterator<String> fieldIterator = jsonForm.fieldNames(); fieldIterator.hasNext(); ) {
+          String fieldName = fieldIterator.next();
+
+          if (!Set.of(fixedFields).contains(fieldName)) {
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartBServiceImpl.java:137]: Same as for PartA

File: src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartBServiceImpl.java [hunk 5]
```diff
@@ -105,4 +113,42 @@ public FormRPartBDto getFormRPartBById(String id, String traineeTisId) {
...
+      String id, String traineeTisId, String[] fixedFields) {
+    log.info("Request to partial delete FormRPartB by id : {}", id);
+
+    try {
+      FormRPartB formRPartB = formRPartBRepository
+          .findByIdAndTraineeTisId(UUID.fromString(id), traineeTisId)
+          .orElse(null);
+
+      if (formRPartB != null) {
+        ObjectMapper objectMapper = new ObjectMapper().findAndRegisterModules();
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/service/impl/FormRPartBServiceImpl.java:131]: Same as for PartA

File: src/test/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListenerTest.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import java.io.IOException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+import uk.nhs.hee.tis.trainee.forms.service.FormRPartAService;
+import uk.nhs.hee.tis.trainee.forms.service.FormRPartBService;
+import uk.nhs.hee.tis.trainee.forms.service.exception.ApplicationException;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListenerTest.java:40]: We're not using any mockito annotations that I could see, see this extension is not required.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (eventDetails[2].equals("formr-a")) {
+          formRPartAService.partialDeleteFormRPartAById(formId, traineeTisId, fixFields);
+          log.info("Partial delete successfully for trainee {} with form Id {} (FormRPartA)",
+              traineeTisId, formId);
+        } else if (eventDetails[2].equals("formr-b")) {
+          formRPartBService.partialDeleteFormRPartBById(formId, traineeTisId, fixFields);
+          log.info("Partial delete successfully for trainee {} with form Id {} (FormRPartB)",
+              traineeTisId, formId);
+        }
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java:77]: We should log if we get an unknown form type too.

Play around and see if a switch case is more readable or not
```java
switch (eventDetails[2]) {
  case "formr-a":
    // Do part a stuff
  case "formr-b":
    // Do part b stuff
  default:
    log.info("unknown form type");
}
````

File: src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+   * Listener for handling form delete form event.
+   */
+  @SqsListener("${application.aws.sqs.delete-event}")
+  public void handleFormDeleteEvent(String message) throws IOException {
+    try {
+      log.info("Form delete event received: {}", message);
+      ObjectMapper objectMapper = new ObjectMapper();
+      DeleteEventDto deleteEvent = objectMapper.readValue(message, DeleteEventDto.class);
+
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java:60]: Formatting: double line break

File: src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+      if (deleteEvent.getDeleteType() == DeleteType.PARTIAL) {
+        final var eventDetails =
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java:63]: Quite a lot of `final var`/`var` usage here that is making it hard to follow outside of an IDE, consider switching to standard type for the declarations unless its something with ridiculous generics all over the place.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * Find all object on S3 and add metadata.
+   */
+  @Execution
+  public void migrate() {
+    log.info("Starting migration to add partial delete related metadata to existing forms on S3.");
+
+    try {
+      List<S3ObjectSummary> keyList = new ArrayList<>();
+      var objects = amazonS3.listObjects(bucketName);
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java:66]: As above, the heavy use of `var` makes this harder to follow than it should be. There is only a single variable with a clear type, I don't know what any of the AWS SDK calls are returning.

File: src/test/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListenerTest.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  @BeforeEach
+  void setUp() {
+    formRPartAService = mock(FormRPartAService.class);
+    formRPartBService = mock(FormRPartBService.class);
+    listener = new FormDeleteEventListener(formRPartAService, formRPartBService);
+  }
+
+  @Test
+  void shouldPartialDeleteFormA() throws IOException {
+    final String MESSAGE = "{\"deleteType\":\"PARTIAL\","
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListenerTest.java:56]: Take advantage of text blocks to make this more readable

e.g.
```java
String message = """
  {
    "deleteType": "PARTIAL",
    "bucket": "document-upload",
    "key": "1/forms/formr-a/1000a.json",
    "fixedFields": ["id", "traineeTisId"]
  }
""";
```

File: src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      for (var obj : keyList) {
+        final var object = amazonS3.getObject(bucketName, obj.getKey());
+
+        var metadata = object.getObjectMetadata();
+        metadata.addUserMetadata("deletetype", DeleteType.PARTIAL.name());
+        metadata.addUserMetadata("fixedfields", FIXED_FIELDS);
+
+        final var request = new PutObjectRequest(
+            bucketName, obj.getKey(), object.getObjectContent(), metadata);
+        amazonS3.putObject(request);
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java:84]: To make this more rerunnable (on failure) you should add a check here for whether the `deleteType`/`fixedFields` metadata already exists. If it does then we skip updating the object.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+      }
+
+      log.info("Finish migration");
+    }
+    catch (Exception e) {
+      log.error("Fail to add metadata to existing forms in bucket " + bucketName + ": ", e);
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java:90]: Given there zero downtime implications for this migrator, I'd maybe let the exception be thrown to stop the migrator.

Already migrated forms should not be reverted, but it gives us a clear indication that we need to fix and re-run rather than having to manually fix afterwards with no option to rerun the migration (without DB fiddling). Combined with the metadata check above we should have a fairly streamlined fail and retry process.

File: src/test/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3Test.java [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+    when(s3.getObject(BUCKET_NAME, "3")).thenReturn(object3);
+
+    migration.migrate();
+
+    verify(s3, times(3)).putObject(putRequestCaptor.capture());
+    List<PutObjectRequest> putObjectRequests = putRequestCaptor.getAllValues();
+    assertThat("Unexpected put object request count.",
+        putObjectRequests.size(), CoreMatchers.is(3));
+
+    var assertId = 1;
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3Test.java:152]: This is an unnecessary use of `var` vs `int`

File: src/test/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3Test.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+          putObjectRequest.getBucketName(), is(BUCKET_NAME));
+      assertThat("Unexpected key.",
+          putObjectRequest.getKey(), is(Integer.toString(assertId)));
+      final var resultInputStream = putObjectRequest.getInputStream();
+      ObjectMapper mapper = new ObjectMapper();
+      Map<String, Object> resultJsonMap = mapper.readValue(resultInputStream, Map.class);
+      assertThat("Unexpected input stream.",
+          resultJsonMap.get("id"), is(Integer.toString(assertId)));
+
+      final var resultUserMetadata =
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3Test.java:164]: As with other `var` comments, it's hard to know what this call returns outside of an IDE

File: src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java [hunk 16]
```diff
@@ -321,4 +332,41 @@ void shouldGetDraftFormRPartAById() {
...
+        DEFAULT_ID_STRING, DEFAULT_TRAINEE_TIS_ID, FIXED_FIELDS);
+
+    verify(repositoryMock).save(formRPartACaptor.capture());
+    FormRPartA dto = formRPartACaptor.getValue();
+    assertThat("Unexpected form ID.", dto.getId(), is(DEFAULT_ID));
+    assertThat("Unexpected trainee ID.", dto.getTraineeTisId(), is(DEFAULT_TRAINEE_TIS_ID));
+    assertThat("Unexpected lifecycle states.",
+        dto.getLifecycleState(), is(LifecycleState.DELETED));
+    assertThat("Unexpected forename.", dto.getForename(), is(nullValue()));
+    assertThat("Unexpected surname.", dto.getSurname(), is(nullValue()));
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java:351]: These should just be `nullValue()` instead of `is(nullValue())`.

File: src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java [hunk 17]
```diff
@@ -321,4 +332,41 @@ void shouldGetDraftFormRPartAById() {
...
+  }
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java:352]: I generally prefer individual field checks (like you've done) over building an "expectedResult" instance of the object being tested, but in this case it might be an appropriate way to test all other fields get nulled without having to test each and every field.

e.g.
```java
FormRPartA expectedDto = new FormRPartA();
expectedDto.setId(DEFAULT_ID);
expectedDto.setTraineeTisId(DEFAULT_TRAINEE_TIS_ID);
expectedDto.setLifeCycle(LifecycleState.DELETED);

assertThat("Unexpected DTO.", dto, is(expectedDto));
```

File: src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java [hunk 18]
```diff
@@ -321,4 +332,41 @@ void shouldGetDraftFormRPartAById() {
...
+
+  @Test
+  void shouldNotPartialDeleteWhenFormRPartANotFoundInDb() {
+    when(repositoryMock.findByIdAndTraineeTisId(DEFAULT_ID, DEFAULT_TRAINEE_TIS_ID))
+        .thenReturn(Optional.ofNullable(null));
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java:357]: Change to return `Optional.empty()` as a cleaner alternative.

File: src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java [hunk 19]
```diff
@@ -321,4 +332,41 @@ void shouldGetDraftFormRPartAById() {
...
+
+    service.partialDeleteFormRPartAById(
+        DEFAULT_ID_STRING, DEFAULT_TRAINEE_TIS_ID, FIXED_FIELDS);
+
+    verify(repositoryMock, never()).save(formRPartACaptor.capture());
+  }
+
+  @Test
+  void shouldThrowExceptionWhenFailToPartialDeleteFormRPartA() throws ApplicationException {
+    doThrow(IllegalArgumentException.class)
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartAServiceImplTest.java:367]: Generally we only need to use this technique for `void` methods, you should be able to still with a more typical
```
when(repositoryMock.findIdAndTraineeTisId(any()), any())).thenThrow(IllegalArgumentException.class);
```

File: src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartBServiceImplTest.java [hunk 20]
```diff
@@ -587,4 +594,65 @@ void shouldGetDraftFormRPartBById() {
         dto.getHavePreviousUnresolvedDeclarations(),
         is(DEFAULT_HAVE_PREVIOUS_UNRESOLVED_DECLARATIONS));
   }
+
+  @Test
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/forms/service/FormRPartBServiceImplTest.java:598]: Same comments as the Part A tests

**Review by Judge40 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java (hunk 1 above)
Comment (DorisWongNHS) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/event/FormDeleteEventListener.java:80]: Yeah, it is just for the edge cases though, trainee forms NON-PARTIAL deletion should not happen (in our current design) after mongock migration, it may be metadata problems / programme bugs that we want to look into. It would be good if we get alerts from DLQ.

**Comment by DorisWongNHS** [comment]
🙈 Oh.. I found that I have misunderstood Sonarcloud, and thought it suggested me to use `var` after Java10. (lol)
https://rules.sonarsource.com/java/RSPEC-6212
Changed it back now.

File: src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java (hunk 12 above)
Comment (DorisWongNHS) [reply src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java]: Good shout. The `addUserMetadata()` would do an "upsert" to update if metadata exists/ insert if not exists. It should keep a correct result even if the migration is rerun though, it's good to skip the unneeded putObject operation if it's already updated.

**Review by Judge40 — APPROVED** [review]
